interfaces/dbus/backend.go

```diff
@@ -132,6 +156,68 @@ func addContent(securityTag string, executableSnippets [][]byte, content map[str
 	}
 }
 
+func (b *Backend) setupBusservices(snapInfo *snap.Info, snippets map[string][][]byte) error {
+	snapName := snapInfo.Name()
+
+	content := map[string]*osutil.FileState{}
+	for _, appInfo := range snapInfo.Apps {
+		securityTag := appInfo.SecurityTag()
+
+		// FIXME: layer violation, we really should do that in the
+		//        dbus interface but without PR #2613 its really tricky
+		for _, slot := range appInfo.Slots {
+			if slot.Interface != "dbus" {
+				continue
+			}
+			bus, ok := slot.Attrs["bus"].(string)
+			if !ok {
+				continue
+			}
+			if bus != "session" {
+				continue
+			}
+			name, ok := slot.Attrs["name"].(string)
+			if !ok {
+				continue
+			}
+			isService := slot.Attrs["service"].(bool)
+			if !isService {
+				continue
+			}
+
+			var buffer bytes.Buffer
+			buffer.Write([]byte(fmt.Sprintf(`[D-BUS Service]
+Name=%s
+Exec=%s
+`, name, appInfo.LauncherCommand())))
+			content[fmt.Sprintf("%s.service", securityTag)] = &osutil.FileState{
+				Content: buffer.Bytes(),
+				Mode:    0644,
+			}
+		}
+	}
+
+	glob := fmt.Sprintf("%s.service", interfaces.SecurityTagGlob(snapName))
+	dir := dirs.SnapDBusSessionServicesFilesDir
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return fmt.Errorf("cannot create directory for DBus service files %q: %s", dir, err)
```

Comment:
The %q/dir is probably redundant here. The error from MkdirAll will include the path name as well.